Amazon.QLDB.Driver.Tests/AsyncQldbSessionTest.cs

```diff
@@ -0,0 +1,266 @@
+        }
+
+        [TestMethod]
+        public void TestAsyncQldbSessionConstrcutorReturnsValidObject()
+        {
+            Assert.IsNotNull(asyncQldbSession);
+        }
+
+        [DataTestMethod]
+        [DynamicData(nameof(CreateExecuteTestData), DynamicDataSourceType.Method)]
+        public async Task ExecuteAsyncTransactionTest(Func<AsyncTransactionExecutor, Task<Object>> transaction, Object expected, Type expectedExceptionType,
+            Type innerExceptionType, Times startTxnTimes, Times executeTimes, Times commitTimes, Times abortTimes, Times retryTimes)
+        {
+            mockSession.Setup(s => s.StartTransactionAsync(It.IsAny<CancellationToken>())).ReturnsAsync(new StartTransactionResult
+            {
+                TransactionId = "testTransactionIdddddd"
+            });
+
+            mockSession.Setup(x => x.ExecuteStatementAsync(
+                It.IsAny<string>(),
+                It.IsAny<string>(),
+                It.IsAny<List<IIonValue>>(),
+                It.IsAny<CancellationToken>())).ReturnsAsync(new ExecuteStatementResult
+                {
+                    FirstPage = new Page
+                    {
+                        NextPageToken = null,
+                        Values = new List<ValueHolder>()
+                    }
+                });
+
+            mockSession.Setup(x => x.CommitTransactionAsync(It.IsAny<string>(), It.IsAny<MemoryStream>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(new CommitTransactionResult
+                {
+                    CommitDigest = new MemoryStream(digest)
+                });
+
+            try
+            {
+                var result = await asyncQldbSession.Execute(transaction);
+
+                Assert.IsNull(expectedExceptionType);
+                Assert.AreEqual(expected, result);
+            }
+            catch (Exception e)
+            {
+                Assert.IsNotNull(expectedExceptionType);
+
+                Assert.IsInstanceOfType(e, expectedExceptionType);
+
+                if (innerExceptionType != null)
+                {
+                    Assert.IsInstanceOfType(e.InnerException, innerExceptionType);
+                }
+            }
+
+            mockSession.Verify(s => s.StartTransactionAsync(It.IsAny<CancellationToken>()), startTxnTimes);
+            mockSession.Verify(s => s.ExecuteStatementAsync(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<List<IIonValue>>(), It.IsAny<CancellationToken>()), executeTimes);
+            mockSession.Verify(s => s.CommitTransactionAsync(It.IsAny<string>(), It.IsAny<MemoryStream>(), It.IsAny<CancellationToken>()), commitTimes);
+
+            mockSession.Verify(s => s.AbortTransactionAsync(It.IsAny<CancellationToken>()), abortTimes);
+            mockAction.Verify(s => s.DisposeDelegate(asyncQldbSession), retryTimes);
+        }
+
+        public static IEnumerable<object[]> CreateExecuteTestData()
```

Comment:
Could this also go in `TestingUtilities.cs`?

Reply:
Yes it can. Same for `CreateExceptions` at the bottom of the file. I will do that in a separate pull request

Reply:
Unfortunately not this one. It has `Func<AsyncTransactionExecutor, Task<object>>` functions which cannot be shared with `QldbSessionTests.cs`